File: Frameworks/SnoizeMIDI/SMPortOutputStream.m
/*
 Copyright (c) 2001-2008, Kurt Revis.  All rights reserved.
 
 Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
 
 * Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
 * Neither the name of Kurt Revis, nor Snoize, nor the names of other contributors may be used to endorse or promote products derived from this software without specific prior written permission.
 
 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */


#import "SMPortOutputStream.h"

#import "SMClient.h"
#import "SMEndpoint.h"
#import "SMHostTime.h"
#import "SMMessage.h"
#import "SMSystemExclusiveMessage.h"
#import "SMSysExSendRequest.h"
#import "SMUtilities.h"


@interface SMPortOutputStream (Private)

- (void)endpointDisappeared:(NSNotification *)notification;
- (void)endpointWasReplaced:(NSNotification *)notification;

- (void)splitMessages:(NSArray *)messages intoCurrentSysex:(NSArray **)sysExMessagesPtr andNormal:(NSArray **)normalMessagesPtr;

- (void)sendSysExMessagesAsynchronously:(NSArray *)sysExMessages;
- (void)sysExSendRequestFinished:(NSNotification *)notification;

@end


@implementation SMPortOutputStream

NSString *SMPortOutputStreamEndpointDisappearedNotification = @"SMPortOutputStreamEndpointDisappearedNotification";
NSString *SMPortOutputStreamWillStartSysExSendNotification = @"SMPortOutputStreamWillStartSysExSendNotification";
NSString *SMPortOutputStreamFinishedSysExSendNotification = @"SMPortOutputStreamFinishedSysExSendNotification";


- (id)init;
{
    OSStatus status;

    if (!(self = [super init]))
        return nil;

    portFlags.sendsSysExAsynchronously = NO;

    sysExSendRequests = [[NSMutableArray alloc] init];
    endpoints = [[NSMutableSet alloc] init];

    status = MIDIOutputPortCreate([[SMClient sharedClient] midiClient], (CFStringRef)@"Output port",  &outputPort);
    if (status != noErr) {
        [NSException raise:NSGenericException format:NSLocalizedStringFromTableInBundle(@"Couldn't create a MIDI output port (error %d)", @"SnoizeMIDI", SMBundleForObject(self), "exception with OSStatus if MIDIOutputPortCreate() fails"), (int)status];
    }

    return self;
}

- (void)dealloc;
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];

    MIDIPortDispose(outputPort);
    outputPort = 0;

    [endpoints release];
    endpoints = nil;

    [sysExSendRequests release];
    sysExSendRequests = nil;    
    
    [super dealloc];
}

- (NSSet *)endpoints;
{
    return endpoints;
}

- (void)setEndpoints:(NSSet *)newEndpoints;
{
    NSNotificationCenter *center;
    NSMutableSet *removedEndpoints, *addedEndpoints;
    NSEnumerator *enumerator;
    SMDestinationEndpoint *endpoint;

    if (!newEndpoints)
        newEndpoints = [NSSet set];

    if ([endpoints isEqual:newEndpoints])
        return;
        
    center = [NSNotificationCenter defaultCenter];

    removedEndpoints = [NSMutableSet setWithSet:endpoints];
    [removedEndpoints minusSet:newEndpoints];
    enumerator = [removedEndpoints objectEnumerator];
    while ((endpoint = [enumerator nextObject]))
        [center removeObserver:self name:nil object:endpoint];

    addedEndpoints = [NSMutableSet setWithSet:newEndpoints];
    [addedEndpoints minusSet:endpoints];
    enumerator = [addedEndpoints objectEnumerator];
    while ((endpoint = [enumerator nextObject])) {
        [center addObserver:self selector:@selector(endpointDisappeared:) name:SMMIDIObjectDisappearedNotification object:endpoint];
        [center addObserver:self selector:@selector(endpointWasReplaced:) name:SMMIDIObjectWasReplacedNotification object:endpoint];
    }
                                
    [endpoints release];
    endpoints = [newEndpoints mutableCopy];        
}

- (BOOL)sendsSysExAsynchronously;
{
    return portFlags.sendsSysExAsynchronously;
}

- (void)setSendsSysExAsynchronously:(BOOL)value;
{
    portFlags.sendsSysExAsynchronously = value;
}

- (void)cancelPendingSysExSendRequests;
{
    [sysExSendRequests makeObjectsPerformSelector:@selector(cancel)];
}

- (NSArray *)pendingSysExSendRequests;
{
    return [NSArray arrayWithArray:sysExSendRequests];
}

//
// SMOutputStream overrides
//

- (void)takeMIDIMessages:(NSArray *)messages;
{
    if ([self sendsSysExAsynchronously]) {
        NSArray *sysExMessages, *normalMessages;

        // Find the messages which are sysex and which have timestamps which are <= now,
        // and send them using MIDISendSysex(). Other messages get sent normally.

        [self splitMessages:messages intoCurrentSysex:&sysExMessages andNormal:&normalMessages];

        [self sendSysExMessagesAsynchronously:sysExMessages];
        [super takeMIDIMessages:normalMessages];
    } else {
        [super takeMIDIMessages:messages];
    }
}

//
// SMOutputStream subclass-implementation methods
//

- (void)sendMIDIPacketList:(MIDIPacketList *)packetList;
{
    NSEnumerator *enumerator;
    SMDestinationEndpoint *endpoint;

    enumerator = [endpoints objectEnumerator];
    while ((endpoint = [enumerator nextObject])) {
        MIDIEndpointRef endpointRef;
        OSStatus status;

        if (!(endpointRef = [endpoint endpointRef]))
            continue;
    
        status = MIDISend(outputPort, endpointRef, packetList);
        if (status) {
#if DEBUG
            NSLog(@"MIDISend(%u, %u, %p) returned error: %ld", (unsigned int)outputPort, (unsigned int)endpointRef, packetList, (long)status);
#endif
        }
    }
}

@end


@implementation SMPortOutputStream (Private)

- (void)endpointDisappeared:(NSNotification *)notification;
{
    SMDestinationEndpoint *endpoint = [notification object];
    NSMutableSet *newEndpoints;

    SMAssert([endpoints containsObject:endpoint]);

    newEndpoints = [NSMutableSet setWithSet:endpoints];
    [newEndpoints removeObject:endpoint];
    [self setEndpoints:newEndpoints];

    [[NSNotificationCenter defaultCenter] postNotificationName:SMPortOutputStreamEndpointDisappearedNotification object:self];
}

- (void)endpointWasReplaced:(NSNotification *)notification;
{
    SMDestinationEndpoint *endpoint = [notification object];
    SMDestinationEndpoint *newEndpoint;
    NSMutableSet *newEndpoints;

    SMAssert([endpoints containsObject:endpoint]);

    newEndpoint = [[notification userInfo] objectForKey:SMMIDIObjectReplacement];

    newEndpoints = [NSMutableSet setWithSet:endpoints];
    [newEndpoints removeObject:endpoint];
    [newEndpoints addObject:newEndpoint];
    [self setEndpoints:newEndpoints];    
}

- (void)splitMessages:(NSArray *)messages intoCurrentSysex:(NSArray **)sysExMessagesPtr andNormal:(NSArray **)normalMessagesPtr;
{
    NSUInteger messageIndex, messageCount;
    NSMutableArray *sysExMessages = nil;
    NSMutableArray *normalMessages = nil;
    MIDITimeStamp now;

    now = SMGetCurrentHostTime();

    messageCount = [messages count];
    for (messageIndex = 0; messageIndex < messageCount; messageIndex++) {
        SMMessage *message;
        NSMutableArray **theArray;

        message = [messages objectAtIndex:messageIndex];
        if ([message isKindOfClass:[SMSystemExclusiveMessage class]] && [message timeStamp] <= now)
            theArray = &sysExMessages;
        else
            theArray = &normalMessages;

        if (*theArray == nil)
            *theArray = [NSMutableArray array];
        [*theArray addObject:message];
    }

    if (sysExMessagesPtr)
        *sysExMessagesPtr = sysExMessages;
    if (normalMessagesPtr)
        *normalMessagesPtr = normalMessages;
}

- (void)sendSysExMessagesAsynchronously:(NSArray *)messages;
{
    NSUInteger messageCount, messageIndex;

    messageCount = [messages count];
    for (messageIndex = 0; messageIndex < messageCount; messageIndex++) {
        SMSystemExclusiveMessage *message;
        NSEnumerator *enumerator;
        SMDestinationEndpoint *endpoint;

        message = [messages objectAtIndex:messageIndex];

        enumerator = [endpoints objectEnumerator];
        while ((endpoint = [enumerator nextObject])) {
            SMSysExSendRequest *sendRequest;

            sendRequest = [SMSysExSendRequest sysExSendRequestWithMessage:message endpoint:endpoint];
            [sysExSendRequests addObject:sendRequest];
            [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(sysExSendRequestFinished:) name:SMSysExSendRequestFinishedNotification object:sendRequest];

            [[NSNotificationCenter defaultCenter] postNotificationName:SMPortOutputStreamWillStartSysExSendNotification object:self userInfo:[NSDictionary dictionaryWithObject:sendRequest forKey:@"sendRequest"]];

            [sendRequest send];
        }
    }
}

- (void)sysExSendRequestFinished:(NSNotification *)notification;
{
    SMSysExSendRequest *sendRequest;

    sendRequest = [notification object];
    SMAssert([sysExSendRequests containsObject:sendRequest]);
    
    [[NSNotificationCenter defaultCenter] removeObserver:self name:nil object:sendRequest];
    [sendRequest retain];
    [sysExSendRequests removeObjectIdenticalTo:sendRequest];

    [[NSNotificationCenter defaultCenter] postNotificationName:SMPortOutputStreamFinishedSysExSendNotification object:self userInfo:[NSDictionary dictionaryWithObject:sendRequest forKey:@"sendRequest"]];

    [sendRequest release];
}

@end

File: Frameworks/SnoizeMIDI/SMPortOutputStream.swift
/*
 Copyright (c) 2001-2021, Kurt Revis.  All rights reserved.

 Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

 * Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
 * Neither the name of Kurt Revis, nor Snoize, nor the names of other contributors may be used to endorse or promote products derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import Foundation
import CoreAudio

@objc public class SMPortOutputStream: SMOutputStream {

    @objc public var endpoints: Set<SMDestinationEndpoint> = [] {
        didSet {
            // The closure-based notification observer API is still awkward to use without creating retain cycles.
            // Easier to use ObjC selectors.
            let center = NotificationCenter.default
            oldValue.subtracting(endpoints).forEach {
                center.removeObserver(self, name: .SMMIDIObjectDisappeared, object: $0)
                center.removeObserver(self, name: .SMMIDIObjectWasReplaced, object: $0)
            }
            endpoints.subtracting(oldValue).forEach {
                center.addObserver(self, selector: #selector(self.endpointDisappeared(notification:)), name: .SMMIDIObjectDisappeared, object: $0)
                center.addObserver(self, selector: #selector(self.endpointWasReplaced(notification:)), name: .SMMIDIObjectWasReplaced, object: $0)
            }
        }
    }

    @objc public var sendsSysExAsynchronously: Bool = false

    @objc public func cancelPendingSysExSendRequests() {
        sysExSendRequests.forEach { _ = $0.cancel() }
    }

    @objc public var pendingSysExSendRequests: [SMSysExSendRequest] {
        return Array(sysExSendRequests)
    }

    @objc public var customSysExBufferSize: Int = 0

    // It's possible (although unlikely) that creating the output port fails.
    // In ObjC we could have made init return nil, but apparently that is not
    // easy to do in Swift (especially to make it usable by ObjC):
    // https://stackoverflow.com/questions/26833845/parameterless-failable-initializer-for-an-nsobject-subclass
    // https://stackoverflow.com/questions/38311365/swift-failable-initializer-init-cannot-override-a-non-failable-initializer
    // Best I can do is to make a new static function that returns a new instance, or nil.
    // Yuck!

    @objc public static func newPortOutputStream() -> SMPortOutputStream? {
        guard let client = SMClient.sharedClient else { return nil }

        var port: MIDIPortRef = 0
        let status = MIDIOutputPortCreate(client.midiClient, "Output Port" as CFString, &port)
        guard status == noErr else { return nil }

        return SMPortOutputStream(port: port)
    }

    private init(port: MIDIPortRef) {
        outputPort = port
        super.init()
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
        MIDIPortDispose(outputPort)
    }

    // MARK: SMOutputStream overrides

    @objc public override func takeMIDIMessages(_ messages: [SMMessage]) {
        if sendsSysExAsynchronously {
            // Find the messages which are sysex and which have timestamps which are <= now,
            // and send them using MIDISendSysex(). Other messages get sent normally.
            let (asyncSysexMessages, normalMessages) = splitMessagesByAsyncSysex(messages)
            sendSysExMessagesAsynchronously(asyncSysexMessages)
            super.takeMIDIMessages(normalMessages)
        }
        else {
            super.takeMIDIMessages(messages)
        }
    }

    // MARK: SMOutputStream subclass-implementation methods

    override func send(_ packetListPtr: UnsafePointer<MIDIPacketList>) {
        for endpoint in endpoints {
            _ = MIDISend(outputPort, endpoint.endpointRef, packetListPtr)
        }
    }

    // MARK: Private

    private var outputPort: MIDIPortRef

    @objc private func endpointDisappeared(notification: Notification) {
        if let endpoint = notification.object as? SMDestinationEndpoint {
            endpointDisappeared(endpoint)
        }
    }

    private func endpointDisappeared(_ endpoint: SMDestinationEndpoint) {
        guard endpoints.contains(endpoint) else { return }
        var newEndpoints = endpoints
        newEndpoints.remove(endpoint)
        endpoints = newEndpoints
        NotificationCenter.default.post(name: .portOutputStreamEndpointDisappeared, object: self)
    }

    @objc private func endpointWasReplaced(notification: Notification) {
        if let endpoint = notification.object as? SMDestinationEndpoint,
           let replacement = notification.userInfo?[SMMIDIObjectReplacement] as? SMDestinationEndpoint {
            endpointWasReplaced(endpoint, replacement)
        }
    }

    private func endpointWasReplaced(_ endpoint: SMDestinationEndpoint, _ replacement: SMDestinationEndpoint) {
        guard endpoints.contains(endpoint) else { return }
        var newEndpoints = endpoints
        newEndpoints.remove(endpoint)
        newEndpoints.insert(replacement)
        endpoints = newEndpoints
    }

    private func splitMessagesByAsyncSysex(_ messages: [SMMessage]) -> ([SMSystemExclusiveMessage], [SMMessage]) {
        // Note: Someday this should use `stablePartition`, when that gets added
        // to the Swift standard library.

        var asyncSysexMessages: [SMSystemExclusiveMessage] = []
        var normalMessages: [SMMessage] = []
        let now = AudioGetCurrentHostTime()

        for message in messages {
            if let sysexMessage = message as? SMSystemExclusiveMessage,
               sysexMessage.timeStamp <= now {
                asyncSysexMessages.append(sysexMessage)
            }
            else {
                normalMessages.append(message)
            }
        }

        return (asyncSysexMessages, normalMessages)
    }

    private var sysExSendRequests = Set<SMSysExSendRequest>()

    private func sendSysExMessagesAsynchronously(_ messages: [SMSystemExclusiveMessage]) {
        let center = NotificationCenter.default

        for message in messages {
            for endpoint in endpoints {
                if let request = SMSysExSendRequest(message: message, endpoint: endpoint, customSysExBufferSize: customSysExBufferSize) {
                    sysExSendRequests.insert(request)

                    var token: NSObjectProtocol?
                    token = center.addObserver(forName: .sysExSendRequestFinished, object: request, queue: nil) { [weak self] _ in
                        guard let self = self else { return }

                        center.removeObserver(token!)
                        token = nil  // Required to break a retain cycle!

                        self.sysExSendRequests.remove(request)

                        center.post(name: .portOutputStreamSysExSendDidEnd, object: self, userInfo: ["sendRequest": request])
                    }

                    center.post(name: .portOutputStreamSysExSendWillBegin, object: self, userInfo: ["sendRequest": request])

                    _ = request.send()
                }
            }
        }
    }

}

// TODO These notifications should just be delegate methods.

public extension Notification.Name {

    static let portOutputStreamEndpointDisappeared = Notification.Name("SMPortOutputStreamEndpointDisappearedNotification")
    // Posted when one of the stream's destination endpoints is removed by the system.

    static let portOutputStreamSysExSendWillBegin = Notification.Name("SMPortOutputStreamWillStartSysExSendNotification")
    static let portOutputStreamSysExSendDidEnd = Notification.Name("SMPortOutputStreamFinishedSysExSendNotification")
    // userInfo has key "sendRequest", object SMSysExSendRequest
    // TODO Formalize that

}

// TODO Duplicate stuff while migrating from ObjC to Swift
@objc public extension NSNotification {

    static let portOutputStreamEndpointDisappeared = Notification.Name.clientSetupChanged
    static let portOutputStreamSysExSendWillBegin = Notification.Name.portOutputStreamSysExSendWillBegin
    static let portOutputStreamSysExSendDidEnd = Notification.Name.portOutputStreamSysExSendDidEnd

}
